### DISCUSSION TITLE
Installing error

### DISCUSSION THREAD

**Discussion by KGhorbani** [pypr/pysph#401] [Q&A]
**Hello,
Thank you for developing this Python package for the SPH method. I tried installing pysph on Windows 10 using "`pip install pysph`", but encountered the following error:**

%%%%%%%%%%%%%%%%%%%%%%
  Using cached PySPH-1.0b1.tar.gz (3.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [69 lines of output]
      C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\Cython\Compiler\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\Users\khash\AppData\Local\Temp\tmpvazn_pot\check_omp.pyx
        tree = Parsing.p_module(s, pxd, full_module_name)
      check_omp.c
         Creating library C:\Users\khash\AppData\Local\Temp\tmpvazn_pot\temp.win-amd64-cpython-39\Release\Users\khash\AppData\Local\Temp\tmpvazn_pot\check_omp.cp39-win_amd64.lib and object C:\Users\khash\AppData\Local\Temp\tmpvazn_pot\temp.win-amd64-cpython-39\Release\Users\khash\AppData\Local\Temp\tmpvazn_pot\check_omp.cp39-win_amd64.exp
      Generating code
      Finished generating code
      warning: pysph\base\nnps_base.pxd:27:13: 'INT_MAX' redeclared
      warning: pysph\base\nnps_base.pxd:27:13: 'INT_MAX' redeclared
      warning: pysph\base\cell_indexing_nnps.pyx:18:0: The 'IF' statement is deprecated and will be removed in a future Cython version. Consider using runtime conditions or C macros instead. See https://github.com/cython/cython/issues/4310

      Error compiling Cython file:
      ------------------------------------------------------------
      ...
          void sort[Iter, Compare](Iter first, Iter last, Compare comp)
          void sort[Iter](Iter first, Iter last)

      IF UNAME_SYSNAME == "Windows":
          @cython.cdivision(True)
          cdef inline double log2(double n) nogil:
               ^
      ------------------------------------------------------------

      pysph\base\cell_indexing_nnps.pyx:20:9: Function signature does not match previous declaration
      ----------------------------------------------------------------------
      Using OpenMP.
      ----------------------------------------------------------------------
      Warning: passing language='c++' to cythonize() is deprecated. Instead, put "# distutils: language=c++" in your .pyx or .pxd file(s)
      Compiling pysph/base/particle_array.pyx because it changed.
      Compiling pysph/base/point.pyx because it changed.
      Compiling pysph/base/nnps_base.pyx because it changed.
      Compiling pysph/base/linked_list_nnps.pyx because it changed.
      Compiling pysph/base/box_sort_nnps.pyx because it changed.
      Compiling pysph/base/spatial_hash_nnps.pyx because it changed.
      Compiling pysph/base/cell_indexing_nnps.pyx because it changed.
      Compiling pysph/base/z_order_nnps.pyx because it changed.
      Compiling pysph/base/stratified_hash_nnps.pyx because it changed.
      Compiling pysph/base/stratified_sfc_nnps.pyx because it changed.
      Compiling pysph/base/octree.pyx because it changed.
      Compiling pysph/base/octree_nnps.pyx because it changed.
      Compiling pysph/base/c_kernels.pyx because it changed.
      Compiling pysph/base/linalg3.pyx because it changed.
      Compiling pysph/tools/mesh_tools.pyx because it changed.
      Compiling pysph/base/gpu_nnps_base.pyx because it changed.
      Compiling pysph/base/z_order_gpu_nnps.pyx because it changed.
      Compiling pysph/base/stratified_sfc_gpu_nnps.pyx because it changed.
      Compiling pysph/base/octree_gpu_nnps.pyx because it changed.
      [ 1/19] Cythonizing pysph/base/box_sort_nnps.pyx
      [ 2/19] Cythonizing pysph/base/c_kernels.pyx
      [ 3/19] Cythonizing pysph/base/cell_indexing_nnps.pyx
      Traceback (most recent call last):
        File "C:\Users\khash\Desktop\PySPH_Project\pysph-venv-309\pysph-env\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Users\khash\Desktop\PySPH_Project\pysph-venv-309\pysph-env\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "C:\Users\khash\Desktop\PySPH_Project\pysph-venv-309\pysph-env\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
        File "C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\setuptools\build_meta.py", line 373, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\setuptools\build_meta.py", line 503, in run_setup
          super().run_setup(setup_script=setup_script)
        File "C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\setuptools\build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 782, in <module>
        File "<string>", line 723, in setup_package
        File "C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\Cython\Build\Dependencies.py", line 1154, in cythonize
          cythonize_one(*args)
        File "C:\Users\khash\AppData\Local\Temp\pip-build-env-tbh84rr4\overlay\Lib\site-packages\Cython\Build\Dependencies.py", line 1321, in cythonize_one
          raise CompileError(None, pyx_file)
      Cython.Compiler.Errors.CompileError: pysph/base/cell_indexing_nnps.pyx
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

%%%%%%%%%%%%%%%%%%%%%%%%%
**I am using Python 3.9.13 and the C++ compiler version is**

Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27051 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

**I have also installed all the required packages as follows** 

Package            Version
------------------ ---------
Beaker             1.13.0
colorama           0.4.6
compyle            0.8.1
cyarray            1.1
Cython             0.29.37
exceptiongroup     1.2.2
importlib_metadata 8.5.0
iniconfig          2.0.0
Mako               1.3.5
MarkupSafe         2.1.5
numpy              2.0.2
packaging          24.1
pip                24.2
platformdirs       4.3.2
pluggy             1.5.0
py-cpuinfo         9.0.0
pygal              3.0.5
pygaljs            1.0.2
pytest             8.3.3
pytest-benchmark   4.0.0
pytools            2024.1.14
setuptools         58.1.0
tomli              2.0.1
typing_extensions  4.12.2
wheel              0.44.0
zipp               3.20.1

**Thank you in advance for your help!**

**Comment by nauaneed** [comment]
There are some fixes that are pending. Meanwhile, could you try this in a fresh python 3.11 venv:

```
pip install --upgrade pip setuptools wheel
pip install numpy cython h5py
pip install https://github.com/pypr/cyarray/zipball/master
pip install https://github.com/pypr/compyle/zipball/master
git clone https://github.com/pypr/pysph.git
cd pysph
git fetch origin pull/400/head:interim
git switch interim
pip install -r requirements.txt -r requirements-test.txt
pip uninstall -y numpy
pip install oldest-supported-numpy
pip install -e . --no-build-isolation
```

I am not sure if it should be done like this, but this might get it going for you until the fixes are done, I guess.

Cheers
Navaneet
